### PR TITLE
AddCompOnMobStateChangeSystem

### LINKED ISSUE #29432: Its too hard to get rid of dead rat king servants
title

### PR DESCRIPTION
**Pull request by lzk228**
## About the PR

Add component and system that will add components on certain mobstate to entity
In this pr used to add ItemComponent to rat servants when they are dead, fixes #29432

## Why / Balance

Good for some mobs

## Technical details

dont think so

## Media

https://github.com/space-wizards/space-station-14/assets/124214523/f465e055-1f25-4b30-9aa7-b08d24c4557a

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- tweak: Rat servants can be picked up when die.

### CODE REVIEW

**Comment by TheShuEd** [comment]
I think it’s better to do it differently.  The rat should always be an item, but the separate "BlockMobStatePickup" component blocks the ability to pickup the entity canceling pickup events

**Comment by metalgearsloth** [comment]
Stop webediting prs or I will start closing them kindly.

File: Content.Shared/Item/MakeItemOnDeadSystem.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            base.Initialize();
+            SubscribeLocalEvent<MakeItemOnDeadComponent, MobStateChangedEvent>(OnMobStateChanged);
+        }
+
+        private void OnMobStateChanged(EntityUid uid, MakeItemOnDeadComponent component, MobStateChangedEvent args)
+        {
+            TryComp<MobStateComponent>(uid, out var mobState);
+
+            if (mobState.CurrentState == MobState.Dead)
```
Comment (metalgearsloth) [inline Content.Shared/Item/MakeItemOnDeadSystem.cs:21]: There's 0 reason this can't use compregistry.

Comment (lzk228) [reply Content.Shared/Item/MakeItemOnDeadSystem.cs:21]: im not sure what do you mean

**Comment by lzk228** [comment]
Did a little refactor, now its more general, not hardcoded to certain variables.

**Comment by lzk228** [comment]
test fail erroneous

File: Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<AddCompOnMobStateChangeComponent, MobStateChangedEvent>(OnMobStateChanged);
+        }
+
+        private void OnMobStateChanged(EntityUid uid, AddCompOnMobStateChangeComponent component, MobStateChangedEvent args)
+        {
+            if(!TryComp<MobStateComponent>(uid, out var mobState))
+            return;
```
Comment (ficcialfaint) [inline Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs:19]:
```suggestion
            if (!TryComp<MobStateComponent>(uid, out var mobState))
                return;
```

File: Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+            if (mobState.CurrentState == component.MobState)
+            {
+                foreach (var compType in component.Components)
+                {
+                    EntityManager.AddComponents(uid, component.Components);
+                }
```
Comment (ficcialfaint) [inline Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs:26]: don't need the foreach loop

File: Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            }
+            else
+            {
+                foreach (var compType in component.Components)
+                {
+                    EntityManager.RemoveComponents(uid, component.Components);
+                }
```
Comment (ficcialfaint) [inline Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs:33]: don't need the foreach loop

File: Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Mobs.Components
+{
```
Comment (ficcialfaint) [inline Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs:4]: use file-scoped namespace

File: Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+using Content.Shared.Mobs.Components;
+
+namespace Content.Shared.Mobs.Systems
+{
```
Comment (ficcialfaint) [inline Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs:4]: use file-scoped namespace

**Review by ficcialfaint — CHANGES_REQUESTED** [review]
also I think would be nice if you could have multiple states with multiple behaviors (e.g. add Comp1 for MobState.Critical and Comp2 for MobState.Dead). Current implementation only limits it to one mobstate if I get it right

File: Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs (hunk 3 above)
Comment (lzk228) [reply Content.Shared/Mobs/Systems/AddCompOnMobStateChangeSystem.cs:26]: oh, i did some changes after adding foreach and forgot to remove it

File: Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+[RegisterComponent]
+public sealed partial class AddCompOnMobStateChangeComponent : Component
+{
+    /// <summary>
+    /// On this state the component will be given.
+    /// </summary>
+    [DataField]
+    public MobState MobState = new();
+
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs:13]: Mobstate is an enum so this constructor is pointless.

File: Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Mobs.Components;
+
+[RegisterComponent]
+public sealed partial class AddCompOnMobStateChangeComponent : Component
+{
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs:7]: Please add access attribute + networked + doc summary as per normal.

File: Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    /// </summary>
+    [DataField]
+    public MobState MobState = new();
+
+    /// <summary>
+    /// Components added to the entity.
+    /// </summary>
+    [DataField]
+    public ComponentRegistry Components = new();
+}
```
Comment (metalgearsloth) [inline Content.Shared/Mobs/Components/AddCompOnMobStateChangeComponent.cs:19]: This should be marked as required.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by lzk228** [comment]
it's a bad implementation, probably should be like MobStateActionsSystem, but i couldn't figure it out with components instead of actions, so closing for now

**Pull request closed without merging**